KEYCLOAK-3142 - Capture ResourceType that triggers an AdminEvent

thomasdarimont opened the pull request:
Introduced new ResourceType enum for AdminEvents which lists
the current supported ResourceTypes for which AdminEvents
can be fired.

Previously it was difficult for custom EventListeners to figure
out which ResourceType triggered an AdminEvent in order
to handle it appropriately, effectively forcing users to parse
the representation.
Having dedicated resource types as a marker on an AdminEvent helps
to ease custom EventListener code.

thomasdarimont committed:
KEYCLOAK-3142 - Capture ResourceType that triggers an AdminEvent

Introduced new ResourceType enum for AdminEvents which lists
the current supported ResourceTypes for which AdminEvents
can be fired.

Previously it was difficult for custom EventListeners to figure
out which ResourceType triggered an AdminEvent in order
to handle it appropriately, effectively forcing users to parse
the representation.
Having dedicated resource types as a marker on an AdminEvent helps
to ease custom EventListener code.

We now also allow filtering of admin events by ResourceType in the
admin-console.

Signed-off-by: Thomas Darimont <thomas.darimont@gmail.com>

thomasdarimont commented on server-spi/src/main/java/org/keycloak/events/admin/ResourceType.java:
```diff
@@ -0,0 +1,155 @@
+ *
+ * @author <a href="mailto:thomas.darimont@gmail.com">Thomas Darimont</a>
+ */
+public enum ResourceType {
```
There is already a similar type in {{org.keycloak.partialimport.ResourceType}} but it is in a seperate package that I didn't want to pull-in. Perhaphs one could unify both enums.

stianst commented:
@mposolda can you review this please?

mposolda commented:
PR looks good to me and will be nice to have ResourceType on admin events. Thanks for the contribution Thomas! 

Just few things:
- Liquibase changelog will need to be updated as "2.0.0" is already released. Sorry the PR was in the queue for longer time :( Also is it possible to remove the precondition for existing column? I suppose you added that because you manually change your DB schema, however each precondition adds additional complexity into changelog and may slow down migration a bit. So if it's possible, can you manually remove the column from your DB before running the migration (or rename column and then add data back to new column after migration is finished if you don't want to loose the info in your DB? I am sure you will figure some way to handle this :) )
- Tests will need to be added. We have AssertAdminEvents to test admin events and it's used on many places. I can see the alternatives like:
1) change all places where AssertAdminEvents are called and add resourceType to them
2) create overriden method on AssertAdminEvents, which won't use resourceType, so you won't need to change all the tests.
The option (1) is more work, but IMO will be better as it will test correct resourceType is on all the places.

thomasdarimont commented:
Hi @mposolda, 

just a quick heads up - I'm applied your suggestions to my local branch and and I'm half-way through with Option 1). I'll try to get it done and pushed this weekend.

Cheers,
Thomas

mposolda commented:
Thanks for the update @thomasdarimont !

thomasdarimont commented:
@mposolda the requested test updates are in place now. I just rebased the branch.
Should be good now.

Cheers,
Thomas

mposolda commented:
@thomasdarimont Thanks for the update. I will followup tomorrow with your PR and will rebase and possibly remove conflicts to have it finally in. However it will be in 2.2.0.CR1 release. Sorry for the delay...

thomasdarimont commented:
Thanks, no worries :-)

mposolda commented:
@thomasdarimont I've tried to followup on your PR, so cherry-picked your commit on top of latest master and sent another PR https://github.com/keycloak/keycloak/pull/3112/commits/f407658e605061a9fb303680b12b58739bb83fa6 . The only thing I changed was to move your stuff to new jpa-changelog-2.2.0 and added that to master changelog. 

However I can see lots of test failures in travis. See https://api.travis-ci.org/jobs/149666747/log.txt?deansi=true . Indeed when I tried to run some tests locally, I figured they are failing.

Will you have time to follow and fix the tests? I know that in your PR, there is easy possibility of conflict. So I can follow if there is few tests failing. However not if almost whole testsuite seems to be failing ;)

thomasdarimont commented:
Thanks @mposolda for having a look - strange, worked for me locally.
How about me rebasing the PR on current master, and moving the changelog stuff to the new 2.2.0 file?
Then I'll have a look at the tests.

Sounds like a plan?

mposolda commented:
@thomasdarimont Sure, you can do either that or just cherry-pick my commit as that's exactly what I did in my commit.

thomasdarimont commented:
Ah I found the reason for the test failures - I didn't implement the asserts for the adminEvents correctly for the client tests. I'll push a the changes as a new commit instead of amending the existing that you can see what I fixed.

thomasdarimont committed:
KEYCLOAK-3142 - Revised according to codereview

Liquibase Moved schema evolution configuration from jpa-changelog-2.1.0
to jpa-changelog-2.2.0.
Corrected wrong ResourceType references in tests.
Adapted AdminEvents copy-routines to be aware of resourceType attribute.
Added ResourceType enum to exposed ENUMS of ServerInfoAdminResource.

Signed-off-by: Thomas Darimont <thomas.darimont@gmail.com>

thomasdarimont commented on services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java:
```diff
@@ -172,7 +172,6 @@ public void deleteRole(final @PathParam("role-name") String roleName) {
             throw new NotFoundException("Could not find client");
         }
 
-        RoleRepresentation rep = getRole(roleName);
```
was not used ... and didn't seem to have a relevant side effect so I removed it.

thomasdarimont commented:
PR should be good now - somehow the travis build kicked in again? ~13min before.

mposolda commented:
yep, I ran it again manually :)

mposolda commented:
@thomasdarimont Thanks for your contribution
